### PR TITLE
Build: allow javac 'release' option, server modules built for Java 11

### LINKED ISSUE #6488: Re-consider Javac target/release versions for all projects
Currently we build _all_ projects for Java 8 (except the Quarkus ones) using the (not recommended) source+target version javac arguments.

There are multiple tasks in this issue:
* Change the "source + target" options to the "release" option. (Verified, that not using the "release" option is not an issue for us at the moment for clients.)
* The former requires a bigger change to `:nessie-client`, which contains source code for Java 11 (Java's new HTTP client). I suspect, the easiest would be to split `:nessie-client` up into `:nessie-client` + new projects `:nessie-client-base` + `:nessie-client.j11` - the first against Java 8, the latter obviously against Java 11.
* Investigate whether we can switch all non-client facing projects (probably everything in `versioned/`, `servers/`, `perftest/`, `gc/`, `compatibility/` + the content-generator tool) to at least Java 11.
* If there would be more Java 11 projects than Java 8 projects, let all projects default to 11 and refactor the current `buildForJava11` to `buildForJava8`.

If we can ensure, that nothing else relies on Java 11 on the server side, we could also bump all server + server-only components to Java 17. However, that J17 bump would leave us with Java 8, Java 11 for the `:nessie-client-j11` and server stuff w/ Java 17.

### PR DESCRIPTION
**Pull request by snazy**
Replaces the `nessie-conventions.gradle.kts` script with a bunch of
specializations for:
* "client" client facing code, for Java 8
* "spark" Nessie Spark code, Java 8
* "iceberg" Nessie components _using_ Iceberg
* "quarkus" Nessie components using Quarkus
* "root" Nessie root projects
* "server" Nessie server components, Java 11

Also removes a bunch of Kotlin files, moving the build code to the
(specialized) call sites, except for `Testing.kt`, which is quite
complex.

Supersedes #6681 + #6682

Fixes #6488

### CODE REVIEW

File: gradle/baselibs.versions.toml [hunk 1]
```diff
@@ -9,6 +9,5 @@ junit-bom = { module = "org.junit:junit-bom", version = "5.9.3" }
 junit-jupiter-api = { module = "org.junit.jupiter:junit-jupiter-api" }
 junit-jupiter-engine = { module = "org.junit.jupiter:junit-jupiter-engine" }
 junit-jupiter-params = { module = "org.junit.jupiter:junit-jupiter-params" }
-nessie-buildsupport-reflectionconfig = { module = "org.projectnessie.buildsupport:reflection-config", version = "0.2.24" }
```
Comment (snazy) [inline gradle/baselibs.versions.toml:12]: Unused in the build-logic, so moved.

File: nessie-iceberg/settings.gradle.kts [hunk 2]
```diff
@@ -16,13 +16,12 @@
 
 import java.util.Properties
 
+includeBuild("../build-logic") { name = "nessie-build-logic" }
+
 if (!JavaVersion.current().isCompatibleWith(JavaVersion.VERSION_11)) {
   throw GradleException("Build requires Java 11")
 }
 
-// Needed by NesQuEIT's manageNessieProjectDependency() in its settings.gradle.kts
```
Comment (snazy) [inline nessie-iceberg/settings.gradle.kts:23]: Appreciated side-effect of the move to `build-logic` is that this hack can disappear.

**Review by adutra — APPROVED** [review]

File: build-logic/src/main/kotlin/nessie-common-src.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+      return configResource
+    }
+  }
+}
+
+// ////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Errorprone
+
+tasks.withType<JavaCompile>().configureEach {
+  if (!project.extra.has("duplicated-project-sources")) {
```
[the remark this answers is not in the document]
Comment (snazy) [reply build-logic/src/main/kotlin/nessie-common-src.gradle.kts:87]: `settings.gradle.kts` (at the bottom)

File: build-logic/src/main/kotlin/nessie-conventions-client.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  id("nessie-java")
+  id("nessie-testing")
+}
+
+tasks.withType<JavaCompile>().configureEach {
+  if (path.startsWith(":nessie-client:")) {
+    // :nessie-client is a little special, because it contains a Java 11 version for
+    // Nessie's dependency-free http client, so setting "release=8", would not work.
+    sourceCompatibility = "1.8"
+    targetCompatibility = "1.8"
```
[the remark this answers is not in the document]
Comment (snazy) [reply build-logic/src/main/kotlin/nessie-conventions-client.gradle.kts:34]: Nope (and we can't actually, because of the Java 11 HTTP client)

**Review by adutra — APPROVED** [review]

**Review by dimas-b — APPROVED** [review]

**Pull request merged**